shell/browser/ui/views/client_frame_view_linux.cc

```diff
@@ -0,0 +1,490 @@
+// Copyright (c) 2021 Ryan Gonzalez.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/ui/views/client_frame_view_linux.h"
+
+#include <algorithm>
+
+#include "base/strings/utf_string_conversions.h"
+#include "cc/paint/paint_filter.h"
+#include "cc/paint/paint_flags.h"
+#include "include/core/SkBlendMode.h"
+#include "include/core/SkColor.h"
+#include "shell/browser/native_window_views.h"
+#include "ui/base/hit_test.h"
+#include "ui/base/l10n/l10n_util.h"
+#include "ui/gfx/canvas.h"
+#include "ui/gfx/font_list.h"
+#include "ui/gfx/geometry/insets.h"
+#include "ui/gfx/geometry/rect.h"
+#include "ui/gfx/skia_util.h"
+#include "ui/gfx/text_constants.h"
+#include "ui/gtk/gtk_compat.h"  // nogncheck
+#include "ui/gtk/gtk_util.h"
+#include "ui/native_theme/native_theme.h"
+#include "ui/strings/grit/ui_strings.h"
+#include "ui/views/controls/button/image_button.h"
+#include "ui/views/linux_ui/linux_ui.h"
+#include "ui/views/linux_ui/nav_button_provider.h"
+#include "ui/views/style/typography.h"
+#include "ui/views/widget/widget.h"
+#include "ui/views/window/frame_buttons.h"
+
+namespace electron {
+
+namespace {
+
+// We can't read border-radius-top-[left/right], so just hardcode Adwaita's
+// value for now.
```

Comment:
> for now

Is this expected to be readable from the system theme somehow in future? I'm not familiar with wayland at all so forgive me if this is a nonsensical question 😅

Reply:
Good point, this is entirely dependent on GTK3 and I don't think they plan on adding an official API for this.

Reply:
Hm, does this mean we'll be locked into not matching the system theme in some cases? Seems like a poor design choice on Wayland's part if that's so, as it doesn't seem possible to blend in.

Reply:
It's definitely rather unfortunate, though it's more on GTK3's part than Wayland's part. There may be increasingly hack-ish ways to derive this value, but I'm just not sure if it would really be worth it: as-is the current choice does seem to blend with the theme choices on the most popular distro variants (matches with the GNOME default Adwaita used on Arch, and looks fine with Yaru used on Ubuntu, I believe the only major distros that ship a theme that conflict would be Pop and *maybe* Manjaro?).

I believe it's worth noting that I believe Chromium itself will just always use a small border radius, though I fear that just using a smaller one will just make it look out of place on all systems.